content/concepts/peer-routing.md

```diff
@@ -3,6 +3,96 @@ title: Peer Routing
+
+Let's see how once can locate in an example. We will call the red node 'Bob', with the prefix `001.` We will call the yellow node Alice, with the prefix `01`
+
+_A brief reminder_: Alice's peerID is a key in the DHT, and the value of this key is Alice's IP address.
+
+Bob can locate Alice in this example. Bob knows Alice's peerID and he needs to get her IP address. From the subtree with the prefix `1` Bob knows a peer with prefix `101`. He sends a message to their address, where he asks for Alice's address.
+
+![1-hop](1-hop.png)
+
+Peer `100` doesn't have Alice's address, so they reroute and ask the peer located at the node with the prefix `110` from their subtree.
+
+![2-hop](2-hop.png)
+
+In this example, peer `110` also doesn't have Alice's address, so they send request to another node, with the prefix `1111`.
+
+![3-hop](3-hop.png)
+
+Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:
+
+![final-hop](final-hop.png)
+
+Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.
+
+### Routing table
+
+The place where one node stores info about other nodes is called a routing table. The routing table is a binary tree whose leaves are `k-buckets`.
+
+### k-buckets
+
+Take a look at the initial tree:
+![Kademlia-system](Kademlia-system.png)
+See these circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
+
+![k-buckets](k-buckets.png)
+
+Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.
+
+## Dynamic DHT
+
+The gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
+
+### Peers Leaving
+
+Let us assume that we have some known parameter `k` that represents some given number of computers. This number should be a value such that it is extremely unlikely that all of those computers will leave the network in the same hour. The original Kademlia paper that the `k` value should be 20.
+
+Next, our `k-bucket` should, instead of storing just 1 computer, store `k` computers within that `k-bucket’s` range. Thus, with high likelihood, there will always be at least one computer online in each of a computer’s `k-buckets`.
+
+And, when hopping, instead of asking one computer from the `k-bucket` we will be asking `k` number of peers. Some may not reply, but not all.
```

Comment:
I believe this isn't all k peers but some other number α of conucrrent reqs we make